docs: start a deployment doc (#181)
Substra/substra-documentation: docs/source/operations/orchestrator/deployment.rst

Deployment
==========

Prerequisites
-------------

In order to deploy the Orchestrator you will need a fully configured Kubernetes cluster.

Preparing your Helm values
--------------------------

The orchestrator deployment is packaged using Helm.
You can find the complete description of values that can be used to configure the chart on `Artifact Hub <https://artifacthub.io/packages/helm/substra/orchestrator>`_.

#. Add the Helm repository for Substra:

   .. code-block:: bash

      helm repo add substra https://substra.github.io/charts/

#. Create a Helm values file named ``orchestrator-values.yaml`` with the following content:

   .. code-block:: yaml

      ingress:
        enabled: true
        hostname: HOSTNAME

   | Replace ``HOSTNAME`` with the hostname of your Orchestrator.
   | This will setup the ingress in order to make your Orchestrator accessible at the defined hostname.

#. Setup your :term:`Substra channels<Channel>`.
   In the ``orchestrator-values.yaml`` file, add the following content:
        
   .. code-block:: yaml

      channels:
        - name: CHANNEL-NAME
          organizations: [ORG-NAME, ...]

>    | Replace ``CHANNEL-NAME`` with the name you want to give to your channel.

Reviewer: are there restrictions in the channel name? e.g. no `-` etc

Author: Not that I am aware of.

>    | Replace ``ORG-NAME, ...`` with the names of the :term:`Organizations<Organization>` that you want to participate in this channel.

Reviewer: ```suggestion
   | Replace ``ORG-NAME, ...`` with the names of the :term:`Organizations<Organization>` that you want to allow to participate in this channel.
```

Author: Do you want to allow them or do you actively want them as part of the channel ?

Reviewer: I'd say you want to allow them. Participating us up to them. But it's very minor, I'll resolve the thread.

>    | If you want more than one channels you can add more items in the list.

Reviewer: ```suggestion
   | If you want more than one channel you can add more items in the list.
```

Here you have created the most basic configuration required to have a running orchestrator that can support a Substra network.
If you want you can jump directly to the section :ref:`deploy-orchestrator` or you can follow along the next sections to enhance the security of your orchestrator.

> Setup TLS for the Orchestrator

Reviewer: redundant 

```suggestion
Setup TLS
```

#. Generate a self-signed Certificate Authority:

   #. Create an openssl config file named ``example-openssl.cnf`` with the following content:

      .. code-block:: ini

         [ req ]
         default_bits		= 2048
         default_md		= sha256
         distinguished_name	= req_distinguished_name

         [ req_distinguished_name ]

         [ v3_ca ]
         basicConstraints = critical,CA:TRUE
         subjectKeyIdentifier = hash
         authorityKeyIdentifier = keyid:always,issuer:always
         keyUsage = cRLSign, keyCertSign

   #. Generate a private key for signing certificates:

      .. code-block:: bash

         openssl genrsa -out orchestrator-ca.key 2048

   #. Generate your Certificate Authority certificate:
        
      .. code-block:: bash

        openssl req -new -x509 -days 365 -sha256 -key orchestrator-ca.key -extensions v3_ca -config example-openssl.cnf -subj "/CN=Orchestrator Root CA" -out orchestrator-ca.crt

#. Generate a certificate for the Orchestrator

   #. Generate a certificate signing request:

      .. code-block:: bash

         openssl req -newkey rsa:2048 -nodes -keyout orchestrator-tls.key -subj "/CN=HOSTNAME" -out orchestrator-cert.csr

>       | Replace ``HOSTNAME`` with the hostname of your Orchestrator as in the ingress configuration.

Reviewer: Slightly clearer

```suggestion
      | Replace ``HOSTNAME`` with the hostname of your Orchestrator. This should be the same HOSTNAME as in the ingress configuration.
```

      .. code-block:: bash

         openssl x509 -req -days 365 -in orchestrator-cert.csr -CA orchestrator-ca.crt -CAkey orchestrator-ca.key -CAcreateserial -out orchestrator-tls.crt -extfile <(printf "subjectAltName=DNS:HOSTNAME")

      | Replace ``HOSTNAME`` with the hostname of your Orchestrator.

      .. caution:: 
         We don't recommend having your certificate valid for a year, you should change this value based on your company policy.

   #. Delete the Certificate Signing Request:

      .. code-block:: bash

         rm orchestrator-cert.csr orchestrator-ca.srl

#. Create a Kubernetes ConfigMap for the CA certificate:
   
   .. code-block:: bash
      
      kubectl create configmap orchestrator-tls-cacert --from-file=orchestrator-ca.crt

#. Create a Kubernetes Secret for the orchestrator TLS key and certificate:

   .. code-block:: bash
      
      kubectl create secret tls orchestrator-tls-server-pair --cert=orchestrator-tls.crt --key=orchestrator-tls.key

.. _deploy-orchestrator:

Deploy the Chart
----------------

To deploy the orchestrator in your Kubernetes cluster follow these steps:

#. Deploy the Orchestrator Helm chart:

   .. code-block:: bash

>       helm install my-orchestrator substra/orchestrator --version 7.4.3 --values orchestrator-values.yaml

Reviewer: :warning: 

```suggestion
      helm install RELEASE-NAME substra/orchestrator --version VERSION --values orchestrator-values.yaml
```

>    | Replace ``RELEASE-NAME`` with the name of your orchestrator release.

Reviewer: slightly clearer 

```suggestion
   | Replace ``RELEASE-NAME`` with the name of your orchestrator release (it can be any arbitrary name)`
```

   .. code-block:: yaml

      ingress:
        enabled: true
        hostname: HOSTNAME

   | Replace ``HOSTNAME`` with the hostname of your Orchestrator.
   | This will setup the ingress in order to make your Orchestrator accessible at the defined hostname.

#. Setup your :term:`Substra channels<Channel>`.
   In the ``orchestrator-values.yaml`` file, add the following content:
        
   .. code-block:: yaml

      channels:
        - name: CHANNEL-NAME
          organizations: [ORG-NAME, ...]

   | Replace ``CHANNEL-NAME`` with the name you want to give to your channel.
   | Replace ``ORG-NAME, ...`` with the names of the :term:`Organizations<Organization>` that you want to participate in this channel.
   | If you want more than one channels you can add more items in the list.

Here you have created the most basic configuration required to have a running orchestrator that can support a Substra network.
If you want you can jump directly to the section deploy or you can follow along the next sections to enhance the security of your orchestrator.

Setup TLS for the Orchestrator
------------------------------

> If you want to you can enable TLS for your orchestrator.

Reviewer 2: We should convey that this is "the right way" and not just a nice thing
```suggestion
In a production environment, it is highly recommended to enable TLS for your orchestrator.
```

> For this you will need to generate a few certificates, here we will generate them manually but you can also use automated tools for this task.

Reviewer 2: ```suggestion
For this, you will need to generate a few certificates. Here we will generate them manually but you can also use automated tools for this task.
```

> If you want to use automated tools we provide a certificate resource for `cert-manager <https://cert-manager.io/>`_, check out the ``orchestrator.tls.createCertificates`` values.

Reviewer 2: Same here, we should highly recommend to use cert-manager. certificates are created with 1y validity, we should probably have a warning that this is the lifetime of the network (unless certs are updated).

To setup TLS, follow these steps:

#. Enable TLS, in the ``orchestrator-values.yaml`` file add the following content:

   .. code-block:: yaml

      orchestrator:
        tls:
          enabled: true

> #. Generate a Certificate Authority certificate:

Reviewer 2: We not only create a certificate here, but also the key:
```suggestion
#. Generate a self-signed Certificate Authority:
```

   #. Create an openssl config file named ``example-openssl.cnf`` with the following content:

      .. code-block:: ini

         [ req ]
         default_bits		= 2048
         default_md		= sha256
         distinguished_name	= req_distinguished_name

         [ req_distinguished_name ]

         [ v3_ca ]
         basicConstraints = critical,CA:TRUE
         subjectKeyIdentifier = hash
         authorityKeyIdentifier = keyid:always,issuer:always
         keyUsage = cRLSign, keyCertSign

   #. Generate a private key for signing certificates:

      .. code-block:: bash

         openssl genrsa -out orchestrator-ca.key 2048

   #. Generate your Certificate Authority certificate:
        
      .. code-block:: bash

>         openssl req -new -x509 -days 365 -sha256 -key orchestrator-ca.key -extensions v3_ca -config example-openssl.cnf -subj "/C=FR/ST=Loire-Atlantique/L=Nantes/O=ORC-ORG-NAME/CN=Orchestrator Root CA" -out orchestrator-ca.crt

Reviewer 2: Can we leave the country, state and location out?

Author: Removed them.

>       | Replace ``ORC-ORG-NAME`` with the name of your organization.

Reviewer 2: What's my organization? Orgs are mentioned on declaring channels but the orc is not really taking part in a specific channel.

Author: Well, I can replace with company name but then what if the user is part of a NGO !

Reviewer 2: I mean, we should explain the meaning and use of the orc organization. As a new user I would certainly ask myself whether it should it be unique or match one of the organization member of a channel?

Author: I removed that part.

      .. code-block:: bash

         openssl x509 -req -days 365 -in orchestrator-cert.csr -CA orchestrator-ca.crt -CAkey orchestrator-ca.key -CAcreateserial -out orchestrator-tls.crt -extfile <(printf "subjectAltName=DNS:HOSTNAME")

      | Replace ``HOSTNAME`` with the hostname of your Orchestrator.

   #. Delete the Certificate Signing Request:

      .. code-block:: bash

         rm orchestrator-cert.csr orchestrator-ca.srl

#. Create a Kubernetes ConfigMap for the CA certificate:
   
   .. code-block:: bash
      
      kubectl create configmap orchestrator-tls-cacert --from-file=orchestrator-ca.crt

#. Create a Kubernetes Secret for the orchestrator TLS key and certificate:

   .. code-block:: bash
      
      kubectl create secret tls orchestrator-tls-server-pair --cert=orchestrator-tls.crt --key=orchestrator-tls.key

We want to validate clients identity usigng mTLS so we need to enable some additional settings:
In the same values file add:

```
orchestrator:
  verifyClientMSPID: true
```

We enable TLS for the server:
in te same values.yaml file
```
orchestrator:
  verifyClientMSPID true
  tls:
    enabled: true
```

> Now for this to work we need to generate some certificates.

Reviewer 2: Can we specify "come certificates"? What for?

Author: This is part of the non-formatted doc.

```

You should now have these files in your current directory: `ca.crt` and `ca.key`.

You can already create a ConfigMap in you cluster named `orchestrator-tls-cacert` using the command:
```
kubectl create configmap orchestrator-tls-cacert --from-file=ca.crt
```

> list configmap to see that it is created

Generate a certificate key for the orchestrator
```
openssl req -newkey rsa:2048 -nodes -keyout orchestrator-tls.key -subj "/C=FR/ST=Loire-Atlantique/L=Nantes/O=Substra/CN=orc.my-corp.org" -out orchestrator-cert.csr
```

and 
```
openssl x509 -req -days 365 -in orchestrator-cert.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out orchestrator-tls.crt -extfile <(printf "subjectAltName=DNS:orc.my-corp.org")
```

Then we need to create a kubernetes secret containing these informations

```
kubectl create secret tls orchestrator-tls-server-pair --cert=orchestrator-tls.crt --key=orchestrator-tls.key
```

Now you can run an orchestrator that is secured

```
helm install my-orchestrator substra/orchestrator --version 7.4.3 --values orchestrator-values.yaml
```

> 

Reviewer 2: ```suggestion
You can check that encrypted communication is possible by listing the different gRPC services:

```

Author: It's the non-formatted copy/paste part of the doc.

```

> GODEBUG=x509sha1=1 grpcurl --cacert ca.crt --rpc-header 'mspid: MyOrg1MSP' --rpc-header 'channel: mychannel' --rpc-header 'chaincode: mycc' orchestrator.install-guide.cg.owkin.tech:443 describe

Reviewer 2: Is this still needed?
```suggestion
grpcurl --cacert ca.crt --rpc-header 'mspid: MyOrg1MSP' --rpc-header 'channel: mychannel' --rpc-header 'chaincode: mycc' orchestrator.install-guide.cg.owkin.tech:443 describe
```

Author: It's the non-formatted copy/paste part of the doc.
